server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -366,6 +401,242 @@ public ClusterState getLastAcceptedState() {
+                // the master service checks if the local node is the master node in order to fail execution of the state update early
+                return clusterStateWithNoMasterBlock(clusterState);
+            }
+            return clusterState;
+        }
+    }
+
+    private ClusterState clusterStateWithNoMasterBlock(ClusterState clusterState) {
+        if (clusterState.nodes().getMasterNodeId() != null) {
+            // remove block if it already exists before adding new one
+            assert clusterState.blocks().hasGlobalBlock(DiscoverySettings.NO_MASTER_BLOCK_ID) == false :
+                "NO_MASTER_BLOCK should only be added by Coordinator";
+            // TODO: allow dynamically configuring NO_MASTER_BLOCK_ALL
+            final ClusterBlocks clusterBlocks = ClusterBlocks.builder().blocks(clusterState.blocks()).addGlobalBlock(
+                DiscoverySettings.NO_MASTER_BLOCK_WRITES).build();
+            final DiscoveryNodes discoveryNodes = new DiscoveryNodes.Builder(clusterState.nodes()).masterNodeId(null).build();
+            return ClusterState.builder(clusterState).blocks(clusterBlocks).nodes(discoveryNodes).build();
+        } else {
+            return clusterState;
+        }
+    }
+
+    @Override
+    public void publish(ClusterChangedEvent clusterChangedEvent, ActionListener<Void> publishListener, AckListener ackListener) {
+        try {
+            synchronized (mutex) {
+                publishUnderLock(clusterChangedEvent, publishListener, ackListener);
+            }
+        } catch (Exception e) {
+            logger.trace(() -> new ParameterizedMessage("[{}] publishing: [{}] failed: {}",
+                getLocalNode().getName(), clusterChangedEvent.source(), e.getMessage()), e);
+            publishListener.onFailure(new FailedToCommitClusterStateException("failure while publishing", e));
+        }
+    }
+
+    private void publishUnderLock(ClusterChangedEvent clusterChangedEvent, ActionListener<Void> completionListener,
+                                  AckListener ackListener) {
+        assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+
+        if (mode != Mode.LEADER) {
+            throw new CoordinationStateRejectedException("publishUnderLock: not currently leading, so cannot handle client value.");
+        }
+
+        if (currentPublication.isPresent()) {
+            throw new CoordinationStateRejectedException("[{}] is in progress", currentPublication.get());
+        }
+
+        // the DiscoveryNodes part can be different as we adapt it in getStateForMasterService when becoming master
+        assert clusterChangedEvent.previousState().term() == coordinationState.get().getLastAcceptedState().term() &&
+            clusterChangedEvent.previousState().version() == coordinationState.get().getLastAcceptedState().version() &&
+            clusterChangedEvent.previousState().metaData() == coordinationState.get().getLastAcceptedState().metaData();
```

Comment:
Do we expect any stronger equality too? For instance if the cluster state has a master node then it should be reference equal to our last-accepted state?

Reply:
I've reworked this in ab43fb15493